File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.cpp
#include <string>
#include <vector>
#include <deque>
#include <iostream>

using namespace std;

class Tree {
public:
    Tree *left = nullptr;
    Tree *right = nullptr;
    bool wolf = false; // true: 늑대, false: 양
};

void traversal(deque<Tree>& q, int& answer, int& wolf, int& sheep) {
    int candidate = q.size();
    while (candidate--) {
        Tree cur = q.front();
        q.pop_front();
        if (cur.wolf) wolf++;
        else sheep++;
        if (wolf < sheep) {
            if (answer < sheep) answer = sheep;
            if (cur.left) q.push_back(*cur.left);
            if (cur.right) q.push_back(*cur.right);
            traversal(q, answer, wolf, sheep);
            if (cur.left) q.pop_back();
            if (cur.right) q.pop_back();
        }
        if (cur.wolf) wolf--;
        else sheep--;
        q.push_back(cur);
    }
}

int solution(vector<int> info, vector<vector<int>> edges) {
    Tree node[17];
    int answer = 0;
    for (int i = 0; i < info.size(); ++i)
        node[i].wolf = info[i] == 1;
    for (vector<int>& edge : edges) {
        Tree& cur = node[edge[0]];
        if (cur.left == nullptr)
            cur.left = &node[edge[1]];
        else
            cur.right = &node[edge[1]];
    }
    deque<Tree> q;
    int wolf = 0, sheep = 0;
    q.push_back(node[0]);
    traversal(q, answer, wolf, sheep);
    return answer;
}

File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.py
from collections import deque
S = 0
W = 1

def is_sheep(node):
    return node[1] == S

def get_children(node):
    return node[2:]

def cnt_sheep(q, tree, sheep, wolf):
    loop = len(q)
    rst = sheep
    for _ in range(loop):
        cur = q.pop()
        if is_sheep(cur):
            new_sheep = sheep + 1
            new_wolf = wolf
        else:
            new_sheep = sheep
            new_wolf = wolf + 1
        children = get_children(cur)
        if new_sheep <= new_wolf:
            q.appendleft(cur)
            continue
        for child in children:
            q.appendleft(tree[child])
        rst = max(rst, cnt_sheep(q, tree, new_sheep, new_wolf))
        for _ in children:
            q.popleft()
        q.appendleft(cur)
    return rst
        
    
def solution(info, edges):
    tree = [[i, j] for i,j in enumerate(info)] # [0]: 0: 양, 1: 늑대 | [1], [2]: 자식노드
    for p, c in edges:
        tree[p].append(c)
    q = deque()
    q.append(tree[0])
    return cnt_sheep(q, tree, 0, 0)
    
    
